### PR TITLE
Beta-Version Enhancement: SOW Versioning & LDAP

### PR DESCRIPTION
**Pull request by ric-evans**
This is the final pull request before gamma testing (with actual institution leaders)

### Versioning Enhancements:
- added date/time for last change on each row, and the person who made the edit
- placed data/time of most recent update to top of page: "Statement[s] of Work as of 05-November 2020 (08:26:53pm CST)"
- added admin-only log/table of each row's changes compared to previous snapshots
- enforced that only explicitly-made snapshots are visible to a non-admin user (a.k.a excel-ingested snapshots are invisible to a non-admin)
    + this goes along with the theme that a snpashot is a version

### LDAP Enhancements:
- users now log in using LDAP credentials -- no more dummy logins
- users select their institution at log-in *(this can eventually be automated once the Keycloak roll-out is complete)*
    + *related: we will need to wait for Keycloak in order to view the institution leaders' names in the WBS summary table*
- **ADMINS:** adding a `~` at the beginning of your username *AND* choosing an institution will launch you into non-admin mode

### CODE REVIEW

File: web_app/data_source/table_config.py [hunk 1]
```diff
@@ -31,17 +29,46 @@ class TableConfigParser:  # pylint: disable=R0904
...
+            self.SOURCE_OF_FUNDS_US_ONLY: Final[str] = "Source of Funds (U.S. Only)"
+            self.FTE: Final[str] = "FTE"
+            self.NSF_MO_CORE: Final[str] = "NSF M&O Core"
+            self.NSF_BASE_GRANTS: Final[str] = "NSF Base Grants"
+            self.US_IN_KIND: Final[str] = "US In-Kind"
+            self.NON_US_IN_KIND: Final[str] = "Non-US In-Kind"
+            self.GRAND_TOTAL: Final[str] = "Grand Total"
+            self.TOTAL_COL: Final[str] = "Total-Row Description"
+            self.TIMESTAMP: Final[str] = "Date & Time of Last Edit"
+            self.EDITOR: Final[str] = "Name of Last Editor"
```
Comment (dsschult) [inline web_app/data_source/table_config.py:54]: I'm unsure why a class of constants needs to be instance variables, and not class variables?  Or why it needs to be a class in the first place, though I can accept that there is a reason for that.

File: web_app/data_source/table_config.py [hunk 2]
```diff
@@ -31,17 +29,46 @@ class TableConfigParser:  # pylint: disable=R0904
...
         self._wbs_l1 = wbs_l1
+
         if cache:
             self._configs = cache
         else:
             self._configs = TableConfigParser.get_configs()
 
+        # set up constants for quick reference
+        self.const = TableConfigParser._Constants()
+        assert all(a in self.get_table_columns() for a in self.const.__dict__.values())
```
Comment (dsschult) [inline web_app/data_source/table_config.py:70]: This better always be true.  asserts are not the easiest things to debug when running, as they don't give good error messages as to what failed.  I usually raise an exception with a proper error message in non-test code.

File: web_app/utils/login.py [hunk 3]
```diff
@@ -1,69 +1,123 @@
...
+
+
+def _ldap_try_login(uid: str, pwd: str) -> None:
+    """Try to get an LDAP connection.
+
+    https://github.com/WIPACrepo/iceprod/blob/master/iceprod/server/rest/auth.py#L454
+    https://www.programcreek.com/python/example/107948/ldap3.Connection
+    """
+    try:
+        _ = ldap3.Connection(
```
Comment (dsschult) [inline web_app/utils/login.py:62]: is the `_ = ` for linting?  normally I'd just remove that if I didn't care about the return value.

File: web_app/data_source/table_config.py (hunk 1 above)
Comment (ric-evans) [reply web_app/data_source/table_config.py:54]: It's a class so I can check easily if column names have changed on the backend when the web app starts up (`assert all(a in self.get_table_columns() for a in self.const.__dict__.values())`). I think that's worth the cost

File: web_app/data_source/table_config.py (hunk 2 above)
Comment (ric-evans) [reply web_app/data_source/table_config.py:70]: _see the comment above_

File: web_app/utils/login.py (hunk 3 above)
Comment (ric-evans) [reply web_app/utils/login.py:62]: that's a habit I've gotten into, I don't know where I picked it up

File: web_app/data_source/table_config.py (hunk 1 above)
Comment (ric-evans) [reply web_app/data_source/table_config.py:54]: I could'be also named this `columns` instead of `const` but that looks very similar to `get_columns()`

**Comment by dsschult** [comment]
I have no technical objections to merging, so will leave it to others on how the app performs.

**Comment by ric-evans** [comment]
> I have no technical objections to merging, so will leave it to others on how the app performs.

Is my LDAP / REST token auth kosher?

**Comment by dsschult** [comment]
> > I have no technical objections to merging, so will leave it to others on how the app performs.
> 
> Is my LDAP and/or REST token auth kosher?

I'd call it so-so.  But it's fine until Keycloak comes online.

File: web_app/data_source/table_config.py (hunk 1 above)
Comment (ric-evans) [reply web_app/data_source/table_config.py:54]: this just blew up one of my unit tests, so I'm moving the assert to an integration test

**Pull request merged**